contracts/cw-ics721-bridge/src/ibc.rs

```diff
@@ -0,0 +1,416 @@
+                contract_addr: env.contract.address.into_string(),
+                msg: to_binary(&ExecuteMsg::BurnEscrowTokens {
+                    channel: ack.original_packet.src.channel_id,
+                    class_id: msg.classId.clone(),
+                    token_ids: msg.tokenIds.clone(),
+                })?,
+                funds: vec![],
+            };
+            let message = SubMsg::reply_always(message, BURN_ESCROW_TOKENS_REPLY_ID);
+            vec![message]
+        } else {
+            vec![]
+        };
+
+        // cw-plus' ics20 implementation doesn't add an ack_success
+        // here. TODO: should we?
+        Ok(IbcBasicResponse::new()
+            .add_submessages(messages)
+            .add_attribute("method", "acknowledge")
+            .add_attribute("sender", msg.sender)
+            .add_attribute("receiver", msg.receiver)
+            .add_attribute("classId", msg.classId)
+            .add_attribute("token_ids", format!("{:?}", msg.tokenIds)))
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_packet_timeout(
+    deps: DepsMut,
+    _env: Env,
+    msg: IbcPacketTimeoutMsg,
+) -> Result<IbcBasicResponse, ContractError> {
+    handle_packet_fail(deps.as_ref(), msg.packet, "timeout")
+}
+
+fn handle_packet_fail(
+    deps: Deps,
+    packet: IbcPacket,
+    error: &str,
+) -> Result<IbcBasicResponse, ContractError> {
+    // On fail, return the NFT from the escrow. We'll only ever handle
+    // our own packets in this method so we use the packet source
+    // which, as a result, is trusted.
+    let escrow_addr = CHANNELS.load(deps.storage, packet.src.channel_id.clone())?;
+
+    // If this deserialization fails something truly strange has
+    // happened. TODO: do we need to handle this? Otherwise, this
+    // method is faliable.
+    let message: NonFungibleTokenPacketData = from_binary(&packet.data)?;
+    let nft_address = CLASS_ID_TO_NFT_CONTRACT.load(deps.storage, message.classId.clone())?;
+
+    let return_nfts = message
+        .tokenIds
+        .iter() // Can't into_iter() here because we use a reference in the closure.
+        .map(|token_id| -> StdResult<SubMsg<Empty>> {
+            let wasm = WasmMsg::Execute {
+                contract_addr: escrow_addr.to_string(),
+                msg: to_binary(&ics_escrow::msg::ExecuteMsg::Withdraw {
+                    nft_address: nft_address.to_string(),
+                    token_id: token_id.clone(),
+                    receiver: message.sender.clone(),
+                })?, // FIXME: how do we handle a failure here?
+                funds: vec![],
+            };
+            // cw-plus' ics20 implementation fail ACKs the fail, so we
+            // do to. More likely than not, more than one of these
+            // will fail if any fail and there are more than one. In
+            // that case, the ack_fail will still get written as it'll
+            // just repeatedly override itself.
```

Comment:
Am quite skeptical that we need to do this. The cw-plus implementation is also inconsistent here. It does not write an ACK success for good ACKs. Probably worth a second opinion here. Link to spec: https://github.com/cosmos/ibc/tree/master/spec/core/ics-004-channel-and-packet-semantics#processing-acknowledgements